core/src/test/java/org/apache/calcite/test/HepPlannerTest.java

```diff
@@ -267,6 +271,38 @@ private void assertIncludesExactlyOnce(String message, String digest, String sub
     assertThat(listener.getApplyTimes() == 1, is(true));
   }
 
+  @Test void testCleanEdges() {
+    // this case is designed to test
+    // CALCITE-4747:https://issues.apache.org/jira/browse/CALCITE-4747
+    // before the improvement, the following optimization will have 7 edges in the graph,
+    // and now there's only 4 edges.
+    HepProgramBuilder programBuilder = HepProgram.builder();
+    programBuilder.addRuleInstance(CoreRules.FILTER_INTO_JOIN);
+
+    final HepTestListener listener = new HepTestListener(0);
+    HepPlanner planner = new HepPlanner(programBuilder.build());
+    planner.addListener(listener);
+
+    final String sql = "select * from emp e "
+        + "join dept d on e.deptno=d.deptno "
+        + "where e.empno>100";
+
+    try {
+      planner.setRoot(tester.convertSqlToRel(sql).rel);
+      planner.findBestExp();
+
+      // the field graph is private, so we use reflect to get it.
+      Field graphField = planner.getClass().getDeclaredField("graph");
+      graphField.setAccessible(true);
```

Comment:
It claims that there are 7 edges before the operation, but I do not see this verified in the test case. 
Can you please verify it?
If it is difficult to do so, I think this test case can be removed. The test case in `DirectedGraphTest` should be enough to cover the case.

Reply:
exactly, there are 4 edges before the operation, but when u execute the optimization, it will add some new edges and cause the edges to grow before this commit, but in this commit, I remove the outdated edges, so the number of edges will keep 4, I think it's better not remove this integrate case, because this case describes the whole optimization, it may mean more than DirectedGraphTest.

Reply:
I see. Thansk for the explanation.
It seems a complicated process is happening behind the scene (the number of edges increases and then decreases).
However, the process is not reflected in the test case, and unknown to readers of the test case.  

In addition, the number 4 seems volatile, as it may change as the result of an irrelevant change. 

So I tend to prefer this test case removed.

Reply:
done